storage/index.go

```diff
@@ -21,61 +21,260 @@
 ) error {
-	return errIndexingNotImplemented
+	d, err := i.doc(namespace, id, tags)
+	if err != nil {
+		fns.OnIndexFinalize()
+		return err
+	}
+
+	i.RLock()
+	if !i.isOpenWithRLock() {
+		i.RUnlock()
+		i.metrics.InsertAfterClose.Inc(1)
+		fns.OnIndexFinalize()
+		return errDbIndexUnableToWriteClosed
+	}
+
+	async := i.insertMode == dbIndexInsertAsync
+	wg, err := i.insertQueue.Insert(d, fns)
+	i.RUnlock()
+
+	if err != nil {
+		fns.OnIndexFinalize()
+		return err
+	}
+
+	// once the write has been queued in the indexInsertQueue, it assumes
+	// responsibility for calling the lifecycle resource hooks.
+
+	if !async {
+		wg.Wait()
+	}
+
+	return nil
 }
 
 func (i *dbIndex) Query(
 	ctx context.Context,
 	query index.Query,
 	opts index.QueryOptions,
 ) (index.QueryResults, error) {
-	return index.QueryResults{}, errIndexingNotImplemented
+	i.RLock()
+	if !i.isOpenWithRLock() {
+		i.RUnlock()
+		return index.QueryResults{}, errDbIndexUnableToQueryClosed
+	}
+
+	// FOLLOWUP(prateek): push down QueryOptions to restrict results
+	iter, err := i.active.segment.Query(query.Query)
+	i.RUnlock()
+	if err != nil {
+		return index.QueryResults{}, err
+	}
+
+	return index.QueryResults{
+		Iterator:   index.NewIterator(iter, i.opts),
+		Exhaustive: true,
+	}, nil
 }
 
-type databaseIndexWriter interface {
-	Write(
-		ctx context.Context,
-		namespace ident.ID,
-		id ident.ID,
-		tags ident.TagIterator,
-	) error
+func (i *dbIndex) isOpenWithRLock() bool {
+	return i.state == dbIndexStateOpen
+
 }
 
-type dbIndexNoOp struct{}
+func (i *dbIndex) Close() error {
+	i.Lock()
+	defer i.Unlock()
+	state := i.state
+	if state != dbIndexStateOpen {
+		return errDbIndexAlreadyClosed
+	}
 
-func (n dbIndexNoOp) Write(context.Context, ident.ID, ident.ID, ident.TagIterator) error {
-	return nil
+	i.state = dbIndexStateClosed
+	return i.insertQueue.Stop()
+}
+
+func (i *dbIndex) doc(ns, id ident.ID, tags ident.Tags) (doc.Document, error) {
+	fields := make([]doc.Field, 0, 1+len(tags))
+	fields = append(fields, doc.Field{
+		Name:      index.ReservedFieldNameNamespace,
+		Value:     i.clone(ns),
+		ValueType: doc.StringValueType,
+	})
+	for j := 0; j < len(tags); j++ {
+		t := tags[j]
+		name := i.clone(t.Name)
+		if bytes.Equal(name, index.ReservedFieldNameNamespace) {
+			return doc.Document{}, errDbIndexUnableToIndexWithReservedFieldName
+		}
+		fields = append(fields, doc.Field{
+			Name:      name,
+			Value:     i.clone(t.Value),
+			ValueType: doc.StringValueType,
+		})
+	}
+	return doc.Document{
+		ID:     i.clone(id),
+		Fields: fields,
+	}, nil
 }
 
-func (n dbIndexNoOp) Query(context.Context, index.Query, index.QueryOptions) (index.QueryResults, error) {
-	return index.QueryResults{}, nil
+// NB(prateek): we take an independent copy of the bytes underlying
+// any ids provided, as we need to maintain the lifecycle of the indexed
+// bytes separately from the rest of the storage subsystem.
+func (i *dbIndex) clone(id ident.ID) []byte {
```

Comment:
Do you have to do this manual clone to prevent sharing pooled byte slices with the index?

Reply:
Kinda -t he storage code finalises IDs when it expires stuff, so we can't rely on the ident passed to the index to be valid for the entirety of the lifecycle of the index. The clone allows us to bypass that restriction. 

Ideally, we'd maintain only a single reference for the ident and do the ref counting correctly to only release once both the storage and index systems were done with it, but that's a much more pervasive change than I wanted to make to the storage side. Did this as a comprise.

Reply:
Fair enough, I'll have to look back at your code to see what you are cloning, but I think the series.ID is never finalized and would probably be safe to share with the index (several parts of the codebase depend on series.ID not being manually finalized)

Reply:
I think you're right but I'd still prefer to explicitly copy. It makes the lifecycle a lot easier to reason about, and avoids issues if we ever do change to start finalizing ids.